crates/daphne/src/pine/flp.rs

```diff
@@ -174,28 +174,76 @@ impl<F: FftFriendlyFieldElement> PineType<F> {
+        let rand_len_per_test = chunk_count(4, self.cfg.dimension);
+        let mut rand = vec![0_u8; rand_len_per_test * NUM_WR_TESTS];
+        xof.fill(&mut rand[..]);
 
         let mut wr_test_results = [F::zero(); NUM_WR_TESTS];
-        let mut i = 0;
-        for wr_test_result in &mut wr_test_results {
-            for x in gradient {
-                // TODO spec: Consider reversing the order in which we read the byte. We can save a
-                // little computation if we can get rid of the subtraction below.
-                let rand_bits = (buf[i >> 3] >> (6 - (i & 7))) & 0b11;
-                match rand_bits {
-                    0b00 => *wr_test_result -= *x,
-                    0b11 => *wr_test_result += *x,
+        for (wr_test_result, rand_per_test) in wr_test_results
+            .iter_mut()
+            .zip(rand.chunks(rand_len_per_test))
+        {
+            for (gradient_chunk, r) in gradient.chunks_exact(4).zip(rand_per_test) {
+                match r & 3 /* 0b11 */ {
+                    0 => *wr_test_result -= gradient_chunk[0],
+                    3 => *wr_test_result += gradient_chunk[0],
+                    _ => (),
+                };
+
+                match r & 12 /* 0b1100 */ {
+                    0 => *wr_test_result -= gradient_chunk[1],
+                    12 => *wr_test_result += gradient_chunk[1],
+                    _ => (),
+                };
+
+                match r & 48 /* 0b110000 */ {
+                    0 => *wr_test_result -= gradient_chunk[2],
+                    48 => *wr_test_result += gradient_chunk[2],
                     _ => (),
                 };
-                i += 2;
+
+                match r & 192 /* 0b11000000 */ {
+                    0 => *wr_test_result -= gradient_chunk[3],
+                    192 => *wr_test_result += gradient_chunk[3],
+                    _ => (),
+                };
+            }
+
+            let gradient_chunk = gradient.chunks_exact(4).remainder();
+            let r = rand_per_test.last().unwrap();
+
+            if gradient_chunk.is_empty() {
+                continue;
             }
+            match r & 3 /* 0b11 */ {
+                    0 => *wr_test_result -= gradient_chunk[0],
+                    3 => *wr_test_result += gradient_chunk[0],
+                    _ => (),
+                };
+
+            if gradient_chunk.len() < 2 {
+                continue;
+            }
+            match r & 12 /* 0b1100 */ {
+                    0 => *wr_test_result -= gradient_chunk[1],
+                    12 => *wr_test_result += gradient_chunk[1],
+                    _ => (),
+                };
+
+            if gradient_chunk.len() < 3 {
+                continue;
+            }
+            match r & 48 /* 0b110000 */ {
+                    0 => *wr_test_result -= gradient_chunk[2],
+                    48 => *wr_test_result += gradient_chunk[2],
+                    _ => (),
+                };
```

Comment:
```suggestion
          for i,x in gradient_chunk.iter().enumerate() {
            match r & 3 /* 0b11 */ {
                    0 => *wr_test_result -= gradient_chunk[i],
                    3 => *wr_test_result += gradient_chunk[i],
                    _ => (),
                };
            r = r >> 2;
          }
```

Reply:
Done. I also added this loop to the `chunks_exact()` loop above: this doesn't seem to impact performance.